File: architecture-skills/design-patterns/examples/abstract-factory/abstract-factory.cs
// Abstract Factory pattern in C#.
//
// Interfaces model products and the factory; in real applications the
// concrete factory is usually chosen once and registered in a dependency
// injection container.

using System;

public interface IButton
{
    string Render();
}

public interface ICheckbox
{
    string Render();
}

public sealed class MacButton : IButton
{
    public string Render() => "( OK )  // rounded macOS button";
}

public sealed class MacCheckbox : ICheckbox
{
    public string Render() => "◉ Remember me  // macOS checkbox";
}

public sealed class WindowsButton : IButton
{
    public string Render() => "[ OK ]  // rectangular Windows button";
}

public sealed class WindowsCheckbox : ICheckbox
{
    public string Render() => "[x] Remember me  // Windows checkbox";
}

/// <summary>One creation method per product kind; products always match.</summary>
public interface IWidgetFactory
{
    IButton CreateButton();
    ICheckbox CreateCheckbox();
}

public sealed class MacFactory : IWidgetFactory
{
    public IButton CreateButton() => new MacButton();
    public ICheckbox CreateCheckbox() => new MacCheckbox();
}

public sealed class WindowsFactory : IWidgetFactory
{
    public IButton CreateButton() => new WindowsButton();
    public ICheckbox CreateCheckbox() => new WindowsCheckbox();
}

public static class Program
{
    // Client code: depends only on the abstract factory and products.
    private static void RenderDialog(IWidgetFactory factory)
    {
        Console.WriteLine(factory.CreateButton().Render());
        Console.WriteLine(factory.CreateCheckbox().Render());
    }

    public static void Main()
    {
        Console.WriteLine("-- macOS theme --");
        RenderDialog(new MacFactory());

        Console.WriteLine("-- Windows theme --");
        RenderDialog(new WindowsFactory());
    }
}

File: architecture-skills/design-patterns/examples/abstract-factory/abstract-factory.js
// Abstract Factory pattern in JavaScript.
//
// Note: with first-class functions, a factory can be a plain object literal
// of constructors — shown here instead of class hierarchies. Duck typing
// replaces the abstract interfaces; each factory returns a matched family.

class MacButton {
  render() {
    return "( OK )  // rounded macOS button";
  }
}

class MacCheckbox {
  render() {
    return "◉ Remember me  // macOS checkbox";
  }
}

class WindowsButton {
  render() {
    return "[ OK ]  // rectangular Windows button";
  }
}

class WindowsCheckbox {
  render() {
    return "[x] Remember me  // Windows checkbox";
  }
}

// Each factory is an object literal with one creation method per product
// kind. Anything with the same shape works — no interface needed.
const macFactory = {
  name: "macOS",
  createButton: () => new MacButton(),
  createCheckbox: () => new MacCheckbox(),
};

const windowsFactory = {
  name: "Windows",
  createButton: () => new WindowsButton(),
  createCheckbox: () => new WindowsCheckbox(),
};

// Client code: depends only on the factory shape, never on a family.
function renderDialog(factory) {
  console.log(factory.createButton().render());
  console.log(factory.createCheckbox().render());
}

for (const factory of [macFactory, windowsFactory]) {
  console.log(`-- ${factory.name} theme --`);
  renderDialog(factory);
}
